dbms/src/Interpreters/CrossJoinProbeHelper.cpp

```diff
@@ -388,44 +389,156 @@ Block crossProbeBlockImpl(
         if (block_full)
             break;
     }
-    probe_process_info.end_row = current_row;
-    probe_process_info.all_rows_joined_finish = (probe_process_info.end_row == block_rows);
+    probe_process_info.updateEndRow<false>(current_row);
     return probe_process_info.result_block_schema.cloneWithColumns(std::move(dst_columns));
 }
+
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, bool has_null_map>
+std::pair<Block, bool> crossProbeBlockShallowCopyRightBlockImpl(
+    ProbeProcessInfo & probe_process_info,
+    const Blocks & right_blocks)
+{
+    static_assert(KIND != ASTTableJoin::Kind::Cross_LeftOuterAnti);
+    assert(probe_process_info.offsets_to_replicate != nullptr);
+
+    size_t num_existing_columns = probe_process_info.block.columns();
+    if constexpr (has_null_map)
+    {
+        /// skip filtered rows, the filtered rows will be handled at the end of this block
+        while (probe_process_info.start_row < probe_process_info.block.rows() && (*probe_process_info.null_map)[probe_process_info.start_row])
+        {
+            ++probe_process_info.start_row;
+        }
+    }
+
+    if (probe_process_info.start_row == probe_process_info.block.rows())
+    {
```

Comment:
Better extract no_matched_rows logic to a function alone.

Reply:
Done